### PR TITLE
End-to-End Variational Network

### PR DESCRIPTION
**Pull request by tullie**
Implementation for "End-to-End Variational Networks for Accelerated MRI Reconstruction" method.

### CODE REVIEW

**Comment by Bala93** [comment]
I couldn't find a paper with title " End-to-End Variational Networks for Accelerated MRI Reconstruction".  Can you share the paper link ? Or is it this paper " Learning a Variational Network for Reconstruction of Accelerated MRI Data ? ". Thank you for the code. Recently, I was looking for code base on variational methods.

File: models/varnet/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+where `CHALLENGE` is either `singlecoil` or `multicoil`. And `MASK_TYPE` is either `random` (for knee)
+or `equispaced` (for brain). Training logs and checkpoints are saved in `experiments/unet` directory.
+
+To run the model on test data:
+```bash
+python models/varnet/varnet.py --mode test --challenge multicoil --data-path DATA --exp var_net --mask-type MASK_TYPE --out-dir reconstructions --checkpoint MODEL
+```
+where `MODEL` is the path to the model checkpoint from `experiments/var_net/version_0/checkpoints/`.
+
+The outputs will be saved to `reconstructions` directory which can be uploaded for submission.
```
Comment (anuroopsriram) [inline models/varnet/README.md:18]: Add a note saying that our paper (https://arxiv.org/abs/2004.06688) should be cited. The format can be similar to https://github.com/facebookresearch/fastMRI#citing

**Comment by anuroopsriram** [comment]
> I couldn't find a paper with title " End-to-End Variational Networks for Accelerated MRI Reconstruction". Can you share the paper link ? Or is it this paper " Learning a Variational Network for Reconstruction of Accelerated MRI Data ? ". Thank you for the code. Recently, I was looking for code base on variational methods.

Here's the paper. We have just put it up on ArXiv.

File: models/varnet/varnet.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+
+        def sens_reduce(x):
+            x = T.ifft2(x)
+            return T.complex_mul(x, T.complex_conj(sens_maps)).sum(dim=1, keepdim=True)
+
+        def soft_dc(x, input):
+            return torch.where(mask, x - ref_kspace, self.zero) * self.dc_weight
+
+        return current_kspace - \
+                soft_dc(current_kspace, input) - \
```
Comment (Bala93) [inline models/varnet/varnet.py:177]: The variable input is not passed to the forward() of the VarNetBlock class. Am I missing something here?

Comment (tullie) [reply models/varnet/varnet.py]: Yeah i'm not sure how this was running considering input wasn't defined (it seemed to work though). Either way I remove input from this function altogether.
